eZ/Publish/Core/Persistence/Legacy/URL/Query/CriterionHandler/VisibleOnly.php

```diff
@@ -0,0 +1,65 @@
+<?php
+
+namespace eZ\Publish\Core\Persistence\Legacy\URL\Query\CriterionHandler;
+
+use eZ\Publish\API\Repository\Values\URL\Query\Criterion;
+use eZ\Publish\Core\Persistence\Legacy\URL\Query\CriteriaConverter;
+use eZ\Publish\Core\Persistence\Legacy\URL\Query\CriterionHandler;
+use eZ\Publish\Core\Persistence\Database\SelectQuery;
+use PDO;
+
+class VisibleOnly implements CriterionHandler
+{
+    /**
+     * {@inheritdoc}
+     */
+    public function accept(Criterion $criterion)
+    {
+        return $criterion instanceof Criterion\VisibleOnly;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function handle(CriteriaConverter $converter, SelectQuery $query, Criterion $criterion)
+    {
+        return $query->expr->in('ezurl.id', $this->getVisibleOnlySubQuery($query));
+    }
+
+    /**
+     * Generate query that selects ids of visible URLs.
+     *
+     * @param \eZ\Publish\Core\Persistence\Database\SelectQuery $query
+     * @return string
+     */
+    protected function getVisibleOnlySubQuery(SelectQuery $query)
+    {
+        // TODO: The following query requires optimization
+        $subSelect = $query->subSelect();
+        $subSelect
+            ->selectDistinct('ezurl_object_link.url_id')
+            ->from('ezurl_object_link')
+            ->innerJoin(
+                'ezcontentobject_attribute',
+                $query->expr->lAnd(
+                    $query->expr->eq('ezurl_object_link.contentobject_attribute_id', 'ezcontentobject_attribute.id'),
+                    $query->expr->eq('ezurl_object_link.contentobject_attribute_version', 'ezcontentobject_attribute.version')
+                )
+            )
+            ->innerJoin(
+                'ezcontentobject_tree',
+                $query->expr->lAnd(
+                    $query->expr->eq('ezcontentobject_tree.contentobject_id', 'ezcontentobject_attribute.contentobject_id'),
+                    $query->expr->eq('ezcontentobject_tree.contentobject_version', 'ezcontentobject_attribute.version')
+                )
+            )
+            ->where(
+                $query->expr->eq(
+                    'ezcontentobject_tree.is_invisible',
+                    $query->bindValue(0, null, PDO::PARAM_INT)
+                )
+            );
+
+        return $subSelect;
```

Comment:
The return Type hint hints string. Which one is correct? :)

Reply:
Fixed in 417ad5386b725bd637cc5bd704c6df3e4ffe6791. It should be `\eZ\Publish\Core\Persistence\Database\SelectQuery`.